core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -425,15 +450,30 @@ protected void responseWithFailure(Throwable t) {
         protected void doRun() throws Exception {
             setPhase(task, "replica");
             assert request.shardId() != null : "request shardId must be set";
+            ReplicaResult result;
             try (Releasable ignored = acquireReplicaOperationLock(request.shardId(), request.primaryTerm())) {
-                shardOperationOnReplica(request);
-                if (logger.isTraceEnabled()) {
-                    logger.trace("action [{}] completed on shard [{}] for request [{}]", transportReplicaAction, request.shardId(),
-                        request);
-                }
+                result = shardOperationOnReplica(request);
             }
-            setPhase(task, "finished");
-            channel.sendResponse(TransportResponse.Empty.INSTANCE);
+            result.respond(new ActionListener<TransportResponse.Empty>() {
+                @Override
+                public void onResponse(Empty response) {
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("action [{}] completed on shard [{}] for request [{}]", transportReplicaAction, request.shardId(),
+                                request);
+                    }
+                    setPhase(task, "finished");
+                    try {
+                        channel.sendResponse(response);
+                    } catch (Exception e) {
+                        onFailure(e);
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    AsyncReplicaAction.this.onFailure(e);
```

Comment:
do we need to honor things like RetryOnReplicaException? if so let's leave a comment. If not I think it will be simple to read if we just call responseWithFailure?